File: Archive/categorized-by-date/21-4-29/287. Find the Duplicate Number/287. Find the Duplicate Number.py
'''
Number: 287
Topic: Find the Duplicate Number
Date: 2021/4/29
Rate: Medium

Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.

There is only one repeated number in nums, return this repeated number.
'''

class Solution:
    def findDuplicate(self, nums: [int]) -> int:
    	hashmap = dict()
    	for num in nums:
    		if num in hashmap:
    			return num
    		else:
    			hashmap[num] = True

'''
grade / pass
tc / O(1)
sc / O(n) or O(1) if in-place hash
'''

File: Archive/categorized-by-date/21-7-8/287. Find the Duplicate Number/287. Find the Duplicate Number.cpp
/*
Number: 287
Topic: Find the Duplicate Number
Date: 2021/7/8
Rate: Medium
https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/

Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.

There is only one repeated number in nums, return this repeated number.

You must solve the problem without modifying the array nums and uses only constant extra space.

Constraints:
* 1 <= n <= 105
* nums.length == n + 1
* 1 <= nums[i] <= n
* All the integers in nums appear only once except for precisely one integer which appears two or more times.

Follow up:

How can we prove that at least one duplicate number must exist in nums?
Can you solve the problem in linear runtime complexity?

*/

class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int low = 1;
        int high = nums.size()-1;
        while (low < high) {
            int mid = low + (high-low)/2;
            int count = 0;
            for (int i = 0; i < nums.size(); i++) {
                if (nums[i] <= mid) count++;
            }
            if (count <= mid)
                low = mid+1;
            else
                high = mid;
        }

        return low;
    }
};